File: init-scripts/admin-pin.js
const { Client, Users, Databases, ID, Query, Permission, Role } = require('node-appwrite');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

async function main() {
  // Load env from root .env
  const envPath = path.join(__dirname, '..', '.env');
  if (fs.existsSync(envPath)) {
    const env = fs.readFileSync(envPath, 'utf8');
    for (const line of env.split('\n')) {
      const match = line.match(/^([A-Z_]+)=(.*)$/);
      if (match) process.env[match[1]] = match[2];
    }
  }

  // Load config
  const configPath = path.join(__dirname, 'config.yml');
  if (!fs.existsSync(configPath)) {
    console.error(`Error: ${configPath} not found`);
    process.exit(1);
  }

  const yaml = fs.readFileSync(configPath, 'utf8');
  const emailMatch = yaml.match(/email:\s*"?([^"\n]+)"?/);
  if (!emailMatch) {
    console.error('Error: master.email not found in config.yml');
    process.exit(1);
  }
  const masterEmail = emailMatch[1].trim();

  console.log('[init] Setting up master admin account...');
  console.log('[init] Email:', masterEmail);

  // Prompt for PIN
  const pin = await promptPin();
  if (!/^\d{4,6}$/.test(pin)) {
    console.error('Error: PIN must be 4-6 digits');
    process.exit(1);
  }

  // Setup Appwrite client
  const client = new Client()
    .setEndpoint(process.env.APPWRITE_ENDPOINT || 'https://sgp.cloud.appwrite.io/v1')
    .setProject(process.env.APPWRITE_PROJECT_ID)
    .setKey(process.env.APPWRITE_API_KEY);

  const users = new Users(client);
  const db = new Databases(client);

  const PROJECT_ID = process.env.APPWRITE_PROJECT_ID;
  const DB_ID = 'ctrader_auth';

  // Find or create user
  let userId = null;
  try {
    const list = await users.list({ queries: [Query.limit(100)] });
    const found = list.users.find(u => u.email === masterEmail);
    if (found) {
      userId = found.$id;
      console.log('[init] Found existing user', userId);
    }
  } catch (e) {
    console.log('[init] User list error:', e.message);
  }

  if (!userId) {
    console.log('[init] Creating Appwrite user for master...');
    const tempPass = crypto.randomBytes(24).toString('base64').replace(/[=+/]/g, '');
    const newUser = await users.createBcryptUser({
      userId: ID.unique(),
      email: masterEmail,
      password: tempPass,
      name: 'admin',
    });
    userId = newUser.$id;
    console.log('[init] Created user', userId);
  }

  // Update label
  console.log('[init] Setting master label...');
  try {
    await users.updateLabels({ userId, labels: ['master'] });
  } catch (e) {
    console.log('[init] Label update warning:', e.message);
  }

  // Hash PIN with scrypt (same format as PIN auth Function)
  const salt = crypto.randomBytes(16).toString('hex');
  const hash = crypto.scryptSync(pin, salt, 64).toString('hex');
  const pinHash = `${salt}:${hash}`;

  // Upsert slave_accounts row
  console.log('[init] Upserting slave_accounts master row...');
  const masterRowId = `master_${userId}`;

  let existingRow = null;
  try {
    const list = await db.listDocuments(DB_ID, 'slave_accounts', [
      Query.equal('appwrite_user_id', userId),
    ]);
    if (list.documents.length > 0) {
      existingRow = list.documents[0];
    }
  } catch (e) {
    console.log('[init] Row lookup error:', e.message);
  }

  if (!existingRow) {
    console.log('[init] Creating master row...');
    try {
      await db.createDocument(DB_ID, 'slave_accounts', masterRowId, {
        appwrite_user_id: userId,
        username: 'admin',
        pin_hash: pinHash,
        role: 'master',
        grant_id: '',
        status: 'active',
        active: true,
        email: masterEmail,
      }, [
        Permission.read(Role.user(userId)),
        Permission.update(Role.user(userId)),
      ]);
      console.log('[init] Created master row', masterRowId);
    } catch (e) {
      console.error('[init] Row creation failed:', e.message);
      process.exit(1);
    }
  } else {
    console.log('[init] Updating master PIN...');
    try {
      await db.updateDocument(DB_ID, 'slave_accounts', existingRow.$id, {
        pin_hash: pinHash,
        active: true,
      });
      console.log('[init] Updated master PIN on row', existingRow.$id);
    } catch (e) {
      console.error('[init] Row update failed:', e.message);
      process.exit(1);
    }
  }

  console.log('');
  console.log('[init] Master admin ready.');
  console.log('       appwrite_user_id:', userId);
  console.log('       username:        admin');
  console.log('       email:           ', masterEmail);
  console.log('');
  console.log('To rotate the PIN, simply rerun this script.');
}

function promptPin() {
  return new Promise((resolve, reject) => {
    process.stdout.write('Enter master PIN (4-6 digits): ');
    process.stdin.setRawMode?.(true);
    process.stdin.resume();
    process.stdin.setEncoding('utf8');

    let pin = '';
    process.stdin.on('data', (ch) => {
      ch = ch.toString();
      if (ch === '\n' || ch === '\r' || ch === '\u0004') {
        process.stdin.setRawMode?.(false);
        process.stdin.pause();
        process.stdout.write('\n');
        resolve(pin);
      } else if (ch === '\u0003') {
        process.stdin.pause();
        reject(new Error('Cancelled'));
      } else {
        pin += ch;
      }
    });
  });
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: init-scripts/admin-pin.py
#!/usr/bin/env python3
"""init-scripts/admin-pin.py — Idempotently create master admin user + PIN row."""

import os
import sys
import getpass
import hashlib
import secrets
from pathlib import Path

from appwrite.client import Client
from appwrite.services.users import Users
from appwrite.services.tables_db import TablesDB
from appwrite.id import ID
from appwrite.query import Query
from appwrite.permission import Permission
from appwrite.role import Role


def load_env():
    """Load .env from project root."""
    root = Path(__file__).parent.parent
    env_file = root / ".env"
    if env_file.exists():
        with open(env_file) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, val = line.split("=", 1)
                    os.environ.setdefault(key, val)


def load_config():
    """Read master.email from config.yml."""
    here = Path(__file__).parent
    config_file = here / "config.yml"
    if not config_file.exists():
        print(f"Error: {config_file} not found. Copy config.example.yml to config.yml.", file=sys.stderr)
        sys.exit(1)

    import yaml
    with open(config_file) as f:
        cfg = yaml.safe_load(f)

    email = cfg.get("master", {}).get("email", "")
    if not email:
        print("Error: master.email missing in config.yml", file=sys.stderr)
        sys.exit(1)
    return email


def hash_pin(pin: str) -> str:
    """scrypt hash (same format as the Node PIN auth Function)."""
    salt = secrets.token_hex(16)
    h = hashlib.scrypt(pin.encode(), salt=salt.encode(), n=2**14, r=8, p=1, dklen=64)
    return f"{salt}:{h.hex()}"


def main():
    load_env()

    master_email = load_config()
    print(f"[init] Setting up master admin account...")
    print(f"[init] Email: {master_email}")

    # Prompt for PIN
    if sys.stdin.isatty():
        pin = getpass.getpass("Enter master PIN (4-6 digits): ")
    else:
        pin = input("Enter master PIN (4-6 digits): ")
    if not pin.isdigit() or not (4 <= len(pin) <= 6):
        print("Error: PIN must be 4-6 digits.", file=sys.stderr)
        sys.exit(1)

    # Connect to Appwrite
    client = Client()
    client.set_endpoint(os.environ["APPWRITE_ENDPOINT"])
    client.set_project(os.environ["APPWRITE_PROJECT_ID"])
    client.set_key(os.environ["APPWRITE_API_KEY"])

    users = Users(client)
    db = TablesDB(client)

    project_id = os.environ["APPWRITE_PROJECT_ID"]
    db_id = "ctrader_auth"

    # Find or create user
    user_id = None
    try:
        user_list = users.list(queries=[Query.limit(100)])
        for u in user_list.users:
            if u.email == master_email:
                user_id = u.id
                print(f"[init] Found existing user {user_id}")
                break
    except Exception as e:
        print(f"[init] User list warning: {e}")

    if not user_id:
        print("[init] Creating Appwrite user for master...")
        temp_pass = secrets.token_urlsafe(32)
        try:
            new_user = users.create_argon2_user(
                user_id=ID.unique(),
                email=master_email,
                password=temp_pass,
                name="admin",
            )
            user_id = new_user["$id"]
            print(f"[init] Created user {user_id}")
        except Exception as e:
            print(f"Error: Failed to create user: {e}", file=sys.stderr)
            sys.exit(1)

    # Set master label
    print("[init] Setting master label...")
    try:
        users.update_labels(user_id=user_id, labels=["master"])
    except Exception as e:
        print(f"[init] Label update warning: {e}")

    # Hash PIN
    pin_hash = hash_pin(pin)
    master_row_id = f"master_{user_id}"

    # Upsert slave_accounts row
    print("[init] Upserting slave_accounts master row...")
    existing = None
    try:
        rows = db.list_rows(db_id, "slave_accounts", [
            Query.equal("appwrite_user_id", user_id),
        ])
        if rows.get("documents"):
            existing = rows["documents"][0]
    except Exception as e:
        print(f"[init] Row lookup warning: {e}")

    if not existing:
        print("[init] Creating master row...")
        try:
            db.create_row(
                db_id,
                "slave_accounts",
                master_row_id,
                {
                    "appwrite_user_id": user_id,
                    "username": "admin",
                    "pin_hash": pin_hash,
                    "role": "master",
                    "grant_id": "",
                    "status": "active",
                    "active": True,
                    "email": master_email,
                },
                permissions=[
                    Permission.read(Role.user(user_id)),
                    Permission.update(Role.user(user_id)),
                ],
            )
            print(f"[init] Created master row {master_row_id}")
        except Exception as e:
            print(f"Error: Row creation failed: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        print("[init] Updating master PIN...")
        try:
            db.update_row(
                db_id,
                "slave_accounts",
                existing["$id"],
                {"pin_hash": pin_hash, "active": True},
            )
            print(f"[init] Updated master PIN on row {existing['$id']}")
        except Exception as e:
            print(f"Error: Row update failed: {e}", file=sys.stderr)
            sys.exit(1)

    print()
    print("[init] Master admin ready.")
    print(f"       appwrite_user_id: {user_id}")
    print(f"       username:         admin")
    print(f"       email:            {master_email}")
    print()
    print("To rotate the PIN, simply rerun this script.")


if __name__ == "__main__":
    main()
